<a href="https://colab.research.google.com/github/alijahwyatt/assignment2/blob/main/Assignment2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment: Data Wrangling and Exploratory Data Analysis
## Do Q1 and Q2, and one other question.
`! git clone https://www.github.com/alijahwyatt/assignment2`

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about? The paper is about tidying and how it's used to make data cleaning easy and effective.
  2. Read the introduction. What is the "tidy data standard" intended to accomplish? To facilitate exploration and analysis of data, and simplify the development of data analysis tools that work together.
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general." The first sentence means if a datasets are tidy, they all share the same similar qualities of a tidy dataset. However, if they are messy, each dataset can have different qualities or things wrong with them that another messy dataset might not have. The second sentence means when looking at dataset, it's usually easy to see what variables are being measured versus the observations gathered. However, in general, defining something as a variable versus an observation can be difficult.
  4. Read Section 2.2. How does Wickham define values, variables, and observations? A value is either a number or string which belongs to a variable and an observation. A variable contains all values that measure the same underlying attribute across units. An observation contains all values measured on the same unit across attributes.
  5. How is "Tidy Data" defined in section 2.3? Tidy Data is defined as a standard way of mapping the meaning of a dataset to its structure.
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset? The 5 most common problems with messy datasets are column headers are values, not variable names, multiple values are stored in one column, variables are stored in both rows and columns, multiple types of observational units are stored in the same table, and a single observational unit is stored in multiple tables. The data in Table 4 are messy because the variables form both columns and rows, and the column headers are values, not variable names. Melting a dataset is turning its columns into rows.
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"? Table 11 is messy because it has variables spread across rows and columns and the element column isn't a variable. Table 12(a) is molten because id, month, and year were molted with element. Table 12(b) is tidy because there is one variable in each column, and each row represents a day's observations.
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling? It's if tidy data is only useful as the tools that work with it, then tidy tools with be linked with tiny data, making it easy to get stuck in a local maxima where independently changing data structures or data tools will not improve the workflow.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [ ]:
! git clone https://www.github.com/alijahwyatt/assignment2
import numpy as np
import pandas as pd


fatal: destination path 'assignment2' already exists and is not an empty directory.


In [ ]:
df = pd.read_csv('./assignment2/data/airbnb_hw.csv')

In [ ]:
var = 'Price'
print(df[var].unique(), '\n')
print(df[var].describe(), '\n')

['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1,000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1,200' '158' '650' '234' '310' '695' '400' '166'
 '119' '62' '168' '340' '479' '43' '395' '144' '52' 

In [ ]:
df[var] = df[var].str.replace(",","") #turned df to string in order to remove commas
df[var] = pd.to_numeric(df[var], errors= 'coerce') #turned object type to int


In [ ]:
print(df[var].describe(), '\n')
df[var+'_nan'] = df[var].isnull() #checks for missing values
print('Total Missings: \n', sum(df[var+ '_nan']), '\n') #prints total num of missings

count    30478.000000
mean       163.589737
std        197.785454
min         10.000000
25%         80.000000
50%        125.000000
75%        195.000000
max      10000.000000
Name: Price, dtype: float64 

Total Missings: 
 0 



In [ ]:
df = pd.read_csv('./assignment2/data/sharks.csv')

<ipython-input-15-352b380f5f54>:1: DtypeWarning: Columns (10,17,18,19,20,21,24) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('./assignment2/data/sharks.csv')


In [ ]:
var2 = 'Type'
print(df[var2].unique(),'\n')
print(df[var2].value_counts(),'\n')
print(df[var2].describe(),'\n')

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg'] 

Unprovoked             4716
Provoked                593
Invalid                 552
Sea Disaster            239
Watercraft              142
Boat                    109
Boating                  92
Questionable             10
Unconfirmed               1
Unverified                1
Under investigation       1
Boatomg                   1
Name: Type, dtype: int64 

count           6457
unique            12
top       Unprovoked
freq            4716
Name: Type, dtype: object 



In [ ]:
df[var2] = df[var2].replace({
    'Invalid' : 'Unconfirmed Cause of Attack', #Attacked but maybe not by shark
    'Under investigation' : 'Unconfirmed Cause of Attack',
    'Unconfirmed' : 'Missing',
    'Questionable': 'Missing',
    'Unverified':'Missing',
    'Boatomg':'Boating',
    'Boat' : 'Boating',
    }) #grouped Questionable, Unconfirmed, Unverified as Missing because of similar nature and Boatomg and Boat as Boating due to similar nature


In [ ]:
print(df[var2].value_counts(), '\n')

Unprovoked                     4716
Provoked                        593
Unconfirmed Cause of Attack     553
Sea Disaster                    239
Boating                         202
Watercraft                      142
Missing                          12
Name: Type, dtype: int64 



In [ ]:
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url,low_memory=False)

In [ ]:
var3 = 'WhetherDefendantWasReleasedPretrial'
print(df[var3].unique(),'\n')
print(df[var3].value_counts(),'\n')
print(df[var3].describe(),'\n')

In [ ]:
df[var3] = df[var3].replace(9, np.nan) #replacing missing values with nan
print(df[var3].value_counts()) #counts values of 1 and 0

In [ ]:
var4 = 'ImposedSentenceAllChargeInContactEvent'
print(df[var4].unique(),'\n')
print(df[var4].value_counts(),'\n')
print(df[var4].describe(),'\n')

In [ ]:
pd.crosstab(df['ImposedSentenceAllChargeInContactEvent'], df['SentenceTypeAllChargesAtConvictionInContactEvent'])

In [ ]:
df.loc[df['SentenceTypeAllChargesAtConvictionInContactEvent']== (4 or 9), 'ImposedSentenceAllChargeInContactEvent'] == np.nan #sets occurances of 4 or 9 to nan

In [ ]:
df[var4] = df[var4].replace(' ',np.nan) #replaces space with nan

**Q3.** This question provides some practice doing exploratory data analysis and visualization.

The "relevant" variables for this question are:
  - `level` - Level of institution (4-year, 2-year)
  - `aid_value` - The average amount of student aid going to undergraduate recipients
  - `control` - Public, Private not-for-profit, Private for-profit
  - `grad_100_value` - percentage of first-time, full-time, degree-seeking undergraduates who complete a degree or certificate program within 100 percent of expected time (bachelor's-seeking group at 4-year institutions)

1. Load the `./data/college_completion.csv` data with Pandas.
2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
3. Cross tabulate `control` and `level`. Describe the patterns you see.
4. For `grad_100_value`, create a histogram, kernel density plot, boxplot, and statistical description.
5. For `grad_100_value`, create a grouped kernel density plot by `control` and by `level`. Describe what you see. Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `grad_100_value` by `level` and `control`. Which institutions appear to have the best graduation rates?
6. Create a new variable, `df['levelXcontrol']=df['level']+', '+df['control']` that interacts level and control. Make a grouped kernel density plot. Which institutions appear to have the best graduation rates?
7. Make a kernel density plot of `aid_value`. Notice that your graph is "bi-modal", having two little peaks that represent locally most common values. Now group your graph by `level` and `control`. What explains the bi-modal nature of the graph? Use `groupby` and `.describe` to make grouped calculations of statistical descriptions of `aid_value` by `level` and `control`.
8. Make a scatterplot of `grad_100_value` by `aid_value`. Describe what you see. Now make the same plot, grouping by `level` and then `control`. Describe what you see. For which kinds of institutions does aid seem to increase graduation rates?

**Q4.** This question uses the Airbnb data to practice making visualizations.

  1. Load the `./data/airbnb_hw.csv` data with Pandas. You should have cleaned the `Price` variable in question 2, and you'll need it later for this question.
  2. What are are the dimensions of the data? How many observations are there? What are the variables included? Use `.head()` to examine the first few rows of data.
  3. Cross tabulate `Room Type` and `Property Type`. What patterns do you see in what kinds of rentals are available? For which kinds of properties are private rooms more common than renting the entire property?
  4. For `Price`, make a histogram, kernel density, box plot, and a statistical description of the variable. Are the data badly scaled? Are there many outliers? Use `log` to transform price into a new variable, `price_log`, and take these steps again.
  5. Make a scatterplot of `price_log` and `Beds`. Describe what you see. Use `.groupby()` to compute a desciption of `Price` conditional on/grouped by the number of beds. Describe any patterns you see in the average price and standard deviation in prices.
  6. Make a scatterplot of `price_log` and `Beds`, but color the graph by `Room Type` and `Property Type`. What patterns do you see? Compute a description of `Price` conditional on `Room Type` and `Property Type`. Which Room Type and Property Type have the highest prices on average? Which have the highest standard deviation? Does the mean or median appear to be a more reliable estimate of central tendency, and explain why?
  7. We've looked a bit at this `price_log` and `Beds` scatterplot. Use seaborn to make a `jointplot` with `kind=hex`. Where are the data actually distributed? How does it affect the way you think about the plots in 5 and 6?

**Q5.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race? Individuals were provided questionnaires with options to select their race
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter? The government can get a better understanding of the nation's diversity. This data is used to ensure fair political representation and can be used to indentify disparities throughout the U.S. Data quality matters because accurate data allows us to make the most helpful decisions based on that data.
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data? The Census included multiple responses for identifying race, allowing for more diverse responses. However, the addition of "Some other race" can lead to underrepresentation of races which weren't answer choices. To reflect the diversity of the population, surveys should allow an option to type what race people identify with. Furthermore, the Census should continue and provide incentives for collecting data in undercounted areas.
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices. They asked questions categorizing the options as male or female. For constructive critism, they could add more options to account for those who don't identify as male or female.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why? Concerns have to be to ensure privacy of individuals. Challenges can include if values are missing, then that could lead to biases when making decisions based on that data. A good practice might be being transparent regarding how missing data is handled, whereas a bad practice might be making assumptions about the missing data that negatively impact people's identities and lead to bias.
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have? I would have concerns about algorithmic bias, especially if the algorithm is trained on biased data. I would also have ethical concerns regarding privacy as people may not want their the algorithm trained on their data without consent.

**Q6.** Open the `./data/CBO_data.pdf` file. This contains tax data for 2019, explaining where the money comes from that the U.S. Federal Government Spends in terms of taxation on individuals/families and payroll taxes (the amount that your employer pays in taxes on your wages).

For some context, the Federal government ultimately spent about $4.4 trillion in 2019, which was 21% of GDP (the total monetary value of all goods and services produced within the United States). Individual Income Taxes is the amount individuals pay on their wages to the Federal government, Corporate Income Taxes is the taxes individuals pay on capital gains from investment when they sell stock or other financial instruments, Payroll Taxes is the tax your employer pays on your wages, Excises and Customs Duties are taxes on goods or services like sin taxes on cigarettes or alcohol, and Estate and Gift Taxes are taxes paid on transfers of wealth to other people.

1. Get the Millions of Families and Billions of Dollars data into a .csv file and load it with Pandas.
2. Create a bar plot of individual income taxes by income decile. Explain what the graph shows. Why are some values negative?
3. Create a bar plot of Total Federal Taxes by income decile. Which deciles are paying net positive amounts, and which are paying net negative amounts?
4. Create a stacked bar plot for which Total Federal Taxes is grouped by Individual Income Taxes, Payroll Taxes, Excises and Customs Duties, and Estate and Gift Taxes. How does the share of taxes paid vary across the adjusted income deciles? (Hint: Are these the kind of data you want to melt?)
5. Below the Total line for Millions of Families and Billions of Dollars, there are data for the richest of the richest families. Plot this alongside the bars for the deciles above the Total line. Describe your results.
6. Get the Percent Distribution data into a .csv file and load it with Pandas. Create a bar graph of Total Federal Taxes by income decile.
7. A tax system is progressive if higher-income and wealthier individuals pay more than lower-income and less wealthy individuals, and it is regressive if the opposite is true. Is the U.S. tax system progressive in terms of amount paid? In terms of the percentage of the overall total?
8. Do the rich pay enough in taxes? Defend your answer.